repo: created a new branch for review

kaustav-goswami opened the pull request:
A setup to do disaggregated memory simulation using gem5 and SST.

kaustav-goswami committed:
repo: created a new branch for review

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

powerjg commented on disaggregated_memory/boards/riscv_main_board.py:
```diff
@@ -0,0 +1,596 @@
+            cache_hierarchy=cache_hierarchy,
+        )
+
+        self.local_memory = local_memory
+        self.remote_memory = remote_memory
```
Why renaming these here? Above you called them `_local_memory` and here it's renamed. Why?

kaustav-goswami replied:
So, I am using `self._localMemory` and `self._remoteMemory` for the init phase. I wanted local_memory and remote_memory conventions to exist, similar to the other board, so I kept them as well. I can get rid of these and test the code.

powerjg commented on disaggregated_memory/boards/riscv_main_board.py:
```diff
@@ -0,0 +1,596 @@
+        # The remote memory starts anywhere after the local memory ends. We
+        # rely on the user to start and end this range.
+        self._remote_mem_ranges = [
+            self.get_remote_memory().get_mem_ports()[0][0]
```
Why `[0][0]`? Should the interface to `get_mem_ports` be changed?

kaustav-goswami replied:
So `get_mem_ports`' returns a tuple of address range and a port, which is directly used while setting up ruby caches. I did not want to create another function just to get the port or the range. I agree `[0][0]` does look hard-coded.

powerjg commented on disaggregated_memory/boards/arm_main_board.py:
```diff
@@ -0,0 +1,469 @@
+    TODO
+    """
+
+    __metaclass__ = ABCMeta
```
Why is this here?

kaustav-goswami replied:
I'll fix the TODO. I copy-pasted the `__metaclass__`. I'll get rid of that.

kaustav-goswami replied:
Done!

powerjg commented on disaggregated_memory/memories/external_remote_memory_v2.py:
```diff
@@ -0,0 +1,215 @@
+from gem5.utils.override import overrides
+
+
+class ExternalRemoteMemoryV2(AbstractMemorySystem):
```
Don't call this V2

kaustav-goswami replied:
That makes sense. I was thinking about it too. Here is the naming convention that I am thinking of.
`ExternalMemory` -> `AbstractMem`
`ExternalRemoteMemory` -> `AbstractMemorySystem`

kaustav-goswami replied:
Done!

powerjg commented on disaggregated_memory/memories/external_remote_memory_v2.py:
```diff
@@ -0,0 +1,215 @@
+        """
+        super().__init__()
+
+        # TODO: marked for deletion
```
Then delete it :)

kaustav-goswami replied:
I'll fix that in the next patch.

kaustav-goswami replied:
Done!

powerjg commented on disaggregated_memory/memories/external_remote_memory_v2.py:
```diff
@@ -0,0 +1,215 @@
+            # by the user.
+            else:
+                # There is no range information provided by the user. Depending
+                # upon the ISA, we have to fix the address.
```
If it's ISA/system dependent, then do it in the Board class, not the memory class!

kaustav-goswami replied:
This is something that we need to discuss. Ranges should be set on the board imo. But we need a way to plug and play ExternalMemory with a range. This is because the remote memory may not be contiguous. So the question is: `Do the board or the memory need to know where the remote memory?`

powerjg commented on disaggregated_memory/memories/external_remote_memory_v2.py:
```diff
@@ -0,0 +1,215 @@
+        # If an XBar is required, it should be added in the connect_things to
+        # avoid initializing an orphan node.
+        raise NotImplementedError("This is a deprecated method for v2")
+        return self._xbar_required
```
delete

kaustav-goswami replied:
I'll fix this in the next patch.

kaustav-goswami replied:
This is completely deleted now! Even the methods are gone.

powerjg commented on disaggregated_memory/memories/external_remote_memory_v2.py:
```diff
@@ -0,0 +1,215 @@
+
+    def get_link_latency(self):
+        raise NotImplementedError("This is a deprecated method for v2")
+        assert self._xbar_required == True
+        return self._link_latency
```
delete

kaustav-goswami replied:
I'll fix this in the next patch.

kaustav-goswami replied:
done!

powerjg commented on disaggregated_memory/configs/riscv-main.py:
```diff
@@ -0,0 +1,293 @@
+1 TIMING CPU cores and executes `STREAM`. The simulation ends when the
+startup is completed successfully.
+
+* This script has to be executed from SST
```
Is this still true?

kaustav-goswami replied:
I'll fix this in the next patch.

kaustav-goswami replied:
Done!

powerjg commented on disaggregated_memory/configs/riscv-main.py:
```diff
@@ -0,0 +1,293 @@
+# commands to execute to run the simulation.
+mount_cmd = ["mount -t sysfs - /sys;", "mount -t proc - /proc;"]
+
+warn("The command list to execute has to be manually set!")
```
BTW, this is the whole point of the "workload" objects!

kaustav-goswami replied:
We can package all the experiments as workload objects. I can spend some time to see how we can use it and customize switch CPUs.

powerjg commented on disaggregated_memory/configs/riscv-main.py:
```diff
@@ -0,0 +1,293 @@
+    "numastat;",
+]
+
+# Since we are using kvm to boot the system, we can boot the system with
+# systemd enabled!
+cmd = (
+    mount_cmd
+    + ["m5 --addr=0x10010000 exit;"]
+    + local_stream
+    + interleave_stream
+    + remote_stream
+    + ["m5 --addr=0x10010000 exit;"]
+)
```
The comment doesn't make sense with the code.

kaustav-goswami replied:
Sorry about that. This comment was only for the ARM board. I'll also fix the sequence of commands here.

kaustav-goswami replied:
Done!

powerjg commented on disaggregated_memory/configs/riscv-main.py:
```diff
@@ -0,0 +1,293 @@
+
+# Here are the different scenarios:
+# no checkpoint, run everything in gem5
+if args.take_ckpt == "True":
```
We can find a better way to communicate "is this executing in SST".

kaustav-goswami replied:
But the user might also run gem5-only experiments, right? Or should we completely remove that option from these runscipts?

powerjg commented on disaggregated_memory/configs/arm-main.py:
```diff
@@ -0,0 +1,299 @@
+# Copyright (c) 2023-24 The Regents of the University of California
```
WHy is this file separate from riscv-main.py? The only difference is the board and the workload. I think you could factor out all of the common code!

kaustav-goswami replied:
I can do that. I'll update this and push this in the next patch.

kaustav-goswami replied:
Actually, we should keep these configs separate imo. Otherwise, the user has to compile ALL/gem5. We can discuss more on this.

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -79,14 +80,22 @@ SSTResponderSubComponent::setOutputStream(SST::Output* output_)
     output = output_;
 }
-
+/*
```
Delete unused code

kaustav-goswami replied:
I'll fix this in the next patch.

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -99,17 +108,68 @@ SSTResponderSubComponent::handleTimingReq(
+        responseReceiver->getBackdoor(data);
+        assert(data->readable());
+
+        // this can be a huge number
```
What does this comment mean???

kaustav-goswami replied:
I'll fix this in the next patch. The comment is not really relevant but I'll try to word this in a better way.

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -238,11 +309,59 @@ SSTResponderSubComponent::handleRecvRespRetry()
         responseQueue.pop();
 }
 
+// void
```
Delete unused code

kaustav-goswami replied:
I'll fix this in the next patch.

kaustav-goswami replied:
Done. Also added a small comment.

powerjg commented on ext/sst/sst_responder_subcomponent.cc:
```diff
@@ -238,11 +309,59 @@ SSTResponderSubComponent::handleRecvRespRetry()
+                addr, data.size(), data);
+            break;
+        }
+        case gem5::MemCmd::ReadReq: {
+            request = new SST::Interfaces::StandardMem::Read(addr, size);
+            break;
```
I'm not sure I buy this is reasonable. Does this ever happen? If so, what is the result?

kaustav-goswami replied:
This is to mirror functional accesses in SST. I'd still need to decode the gem5 packet so that I can send corresponding timing requests to SST's memory.

powerjg commented on ext/sst/sst_responder_subcomponent.hh:
```diff
@@ -45,19 +45,26 @@
+    // gem5::OutgoingRequestBridge* responseReceiver;
+    // responseReceiver for this branch is hardcoded to ExternalMemory*.
+    // TODO: We need to make a better design to handle multiple types of
+    // outgoing request classes.
```
This must be done sooner rather than later. We can chat about the best approach.

kaustav-goswami replied:
I agree. I ran into this issue yesterday so I tried to solve it hastily.

powerjg reviewed (commented):
Overall, this looks very good! Great job cleaning things up.

Let's focus on the experiments now, and we can clean things up futher soon.

Let's focus on writing a paper on this and improving the code so we can release it.

kaustav-goswami committed:
ext,config,mem: updated the code for review

This change addresses most of the comments from the PR.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
ext,doc: added a documentation and sst-script

This change adds a document to explain how this system works and how a
user can start simulations using this platform. In addition, this change
also adds a SST-side runscript to start the phase 2 simulations.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs,ext: final touchups for the composable project

This change includes the following:
- Checkpoint restore/base clock was missing in the previous update
- The L3 crossbar bug was fixed by increasing the size of the snoop
  filter.
- Remote link is now deprecated. A fatal message is now displayed.
- Remote memory connection with the membus is moved back to the
  cachehierarchies.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs: hotfix to include fatal

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
doc: updated the documentation

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs: upstreamed experiment scripts (partial)

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
doc: updated the README wtih traffic generator instructions

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
exp: starting to prepare for the experiments

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
exp: starting to prepare for the experiments

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs: fixed the system; added an npb script

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
configs: fixed the system; added an npb script

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

mbabaie committed:
fixed the issues in the main scripts

kaustav-goswami committed:
temp: uploading to kg to fix ext mem

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
stats,ext: added new stats in the external memory

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

mbabaie committed:
shrank the scripts and puut everything in the same dir: disaggdisaggregated_memory

mbabaie committed:
merging with latest changes from the repo

mbabaie committed:
all the shrunk scripts for running stream and npb, with checkpoint or restore in their name, and in SST folder

kaustav-goswami committed:
ext,mem: Fixed incorrect bandwidth calculation and added a new stat to count outstanding packets

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

mbabaie committed:
fixed the remote memory ranges for np apps and added md5sum for resources

mbabaie committed:
Merge branch 'kg/composable-memory-2' of https://github.com/darchr/gem5 into kg/composable-memory-2

mbabaie committed:
latest version of scripts for experiments

kaustav-goswami committed:
x86,ext: added a composable memory board to the repo

This change adds a X86 Composable Memory board to the repository. A
sample script is also added. The user needs to fake NUMA information to
the OS/board to enable gem5 + SSt simulations.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
config,ext: added a arm shared memory board

This change adds a new ARM shared memory board, which allows users to
perform shared memory programs across two independent systems. The board
uses a UIO device to expose a physical memory address range to the OS,
which can be then mmaped by a C program in each of the systems.

Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>

kaustav-goswami committed:
arch-x86,stdlib: updated the x86 main board with madt

This change fixes the kernel bug on startup and adds madt entries to the
X86 board.

Change-Id: Idd60cc485f8b38628d10998036f034e519eb11e6
Signed-off-by: Kaustav Goswami <kggoswami@ucdavis.edu>
